[NCCL] Support Wait Timeout in ProcessGroupNCCL

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #41265 [Gloo] Tests for Gloo Async Work Wait-level Timeouts
* #40948 [Gloo] Support work-level timeouts in ProcessGroupGloo
* #40947 [NCCL] Tests for WorkNCCL::wait with Timeouts
* **#40946 [NCCL] Support Wait Timeout in ProcessGroupNCCL**
* #40944 [NCCL] Add timeout to ProcessGroup Work Wait

Adds timeout to ProcessGroupNCCL::wait. Currently, WorkNCCL objects already have a timeout set during ProcessGroupNCCL construction. The new wait function will override the existing timeout with the user-defined timeout if one is provided. Timed out operations result in NCCL communicators being aborted and an exception being thrown.

Differential Revision: [D22127898](https://our.internmc.facebook.com/intern/diff/D22127898/)

osalpekar committed:
[NCCL] Support Wait Timeout in ProcessGroupNCCL

Adds timeout to ProcessGroupNCCL::wait. Currently, WorkNCCL objects already have a timeout set during ProcessGroupNCCL construction. The new wait function will override the existing timeout with the user-defined timeout if one is provided. Timed out operations result in NCCL communicators being aborted and an exception being thrown.

Differential Revision: [D22127898](https://our.internmc.facebook.com/intern/diff/D22127898/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Support Wait Timeout in ProcessGroupNCCL"


Adds timeout to ProcessGroupNCCL::wait. Currently, WorkNCCL objects already have a timeout set during ProcessGroupNCCL construction. The new wait function will override the existing timeout with the user-defined timeout if one is provided. Timed out operations result in NCCL communicators being aborted and an exception being thrown.

Differential Revision: [D22127898](https://our.internmc.facebook.com/intern/diff/D22127898/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Support Wait Timeout in ProcessGroupNCCL"


Adds timeout to ProcessGroupNCCL::wait. Currently, WorkNCCL objects already have a timeout set during ProcessGroupNCCL construction. The new wait function will override the existing timeout with the user-defined timeout if one is provided. Timed out operations result in NCCL communicators being aborted and an exception being thrown.

Differential Revision: [D22127898](https://our.internmc.facebook.com/intern/diff/D22127898/)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -68,7 +68,7 @@ class ProcessGroupNCCL : public ProcessGroup {
 
     // Same as calling synchronize() for NCCL work.
-    bool wait(std::chrono::milliseconds timeout = kUnsetTimeout) override;
+    bool wait(std::chrono::milliseconds timeout = kNoTimeout) override;
```
Any reason for using default timeout value for `wait`, but explicitly avoiding introduce an argument to `ProcessGroupNCCL::WorkNCCL::synchronize()`? Will default value also work for `synchronize()`?

osalpekar replied:
Briefly chatted with Hongyi about this - we were mainly aiming to change the `wait` API only without many other side-effects on user facing functions, so decided to keep `synchronize` as is. If you think it's better to have a timeout with default value for `synchronize`, we can do that as well.

osalpekar committed:
Update on "[NCCL] Support Wait Timeout in ProcessGroupNCCL"


Adds timeout to ProcessGroupNCCL::wait. Currently, WorkNCCL objects already have a timeout set during ProcessGroupNCCL construction. The new wait function will override the existing timeout with the user-defined timeout if one is provided. Timed out operations result in NCCL communicators being aborted and an exception being thrown.

Differential Revision: [D22127898](https://our.internmc.facebook.com/intern/diff/D22127898/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Support Wait Timeout in ProcessGroupNCCL"


Adds timeout to ProcessGroupNCCL::wait. Currently, WorkNCCL objects already have a timeout set during ProcessGroupNCCL construction. The new wait function will override the existing timeout with the user-defined timeout if one is provided. Timed out operations result in NCCL communicators being aborted and an exception being thrown.

Differential Revision: [D22127898](https://our.internmc.facebook.com/intern/diff/D22127898/)

[ghstack-poisoned]

jiayisuse commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -202,8 +202,15 @@ void ProcessGroupNCCL::WorkNCCL::checkAndThrowException() {
 void ProcessGroupNCCL::WorkNCCL::synchronize() {
+  // Call Synchronize without a timeout. We use this method to avoid adding a
+  // timeout argument to the public synchronize API.
+  synchronizeInternal(kNoTimeout);
```
Thanks for making synchronizeInternal() accepts explicit timeout

osalpekar committed:
Update on "[NCCL] Support Wait Timeout in ProcessGroupNCCL"


Adds timeout to ProcessGroupNCCL::wait. Currently, WorkNCCL objects already have a timeout set during ProcessGroupNCCL construction. The new wait function will override the existing timeout with the user-defined timeout if one is provided. Timed out operations result in NCCL communicators being aborted and an exception being thrown.

Differential Revision: [D22127898](https://our.internmc.facebook.com/intern/diff/D22127898/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Support Wait Timeout in ProcessGroupNCCL"


Adds timeout to ProcessGroupNCCL::wait. Currently, WorkNCCL objects already have a timeout set during ProcessGroupNCCL construction. The new wait function will override the existing timeout with the user-defined timeout if one is provided. Timed out operations result in NCCL communicators being aborted and an exception being thrown.

Differential Revision: [D22127898](https://our.internmc.facebook.com/intern/diff/D22127898/)

[ghstack-poisoned]
